pycrumbs/__init__.py

```diff
@@ -0,0 +1,352 @@
+    root = _build(entities, 0, None)
+    for idx, node in enumerate(root.walk()):
+        node.uid = idx
+
+    return root
+
+
+def collapse(node, min_entities_per_node):
+    """\
+    Collapses small nodes into a single "other" node at each level. This operation creates a copy
+    of the tree.
+
+    :param node: root node/subtree
+    :param min_entities_per_node: minimum number of entities per node
+    :return: transformed tree (new object)
+
+    """
+    if not node.children:
+        return node
+
+    new_node = Node(node.trajectories, [],
+                    node.entities, node.name, node.uid, node.parent)
+
+    small_children = [child for child in node.children if len(child.entities) < min_entities_per_node]
+    large_children = [child for child in node.children if len(child.entities) >= min_entities_per_node]
+
+    if not large_children:  # collapsing would create a single child, so get rid of children altogether
+        return new_node
+
+    new_children = [collapse(child, min_entities_per_node) for child in large_children]
+    if small_children:
+        other_trajectories = set()
+        other_entities = []
+        for child in small_children:
+            other_trajectories.update(child.trajectories)
+            other_entities += child.entities
+
+        new_children.append(Node(other_trajectories, [], other_entities, name="Other",
+                                 uid=small_children[0].uid, parent=new_node))
+    new_node.children = new_children
+
+    return new_node
+
+
+def pretty_format_tree(root, indent=4):
+    """\
+    Pretty formats a tree, returning a string
+
+    :param root: root of the subtree to format
+    :param indent: indentation between levels (default 4)
+    :return: pretty tree as string
+
+    """
+    def _format(node, parent, current_indent):
+        txt = ""
+        txt += "{}* n={} ({:.1%}). Trajectories: {}\n".format(" " * current_indent, len(node.entities),
+                                                              1.0 * len(node.entities) / len(parent.entities),
+                                                              ", ".join(str(traj) for traj in node.trajectories))
+
+        txt += "".join(_format(child, node, current_indent + indent)
+                       for child in node.children)
+        return txt
+
+    return _format(root, root, 0).strip()
+
+
+def new_observation(parent, child, max_observations=1):
+    """\
+    Gets the new observations between a parent and its children
+    :param parent: parent node
+    :param child: child node
+    :param max_observations: for nodes with multiple associated trajectories, the new "observation" will be a union
+    e.g. "OBS1 or OBS2 or OBS3". max_observations specifies the max number of alternatives before the result
+    is abbreviated as "other". Default 1
+
+    """
+    def begins_with(lst, prefix):
+        """True iff lst begins with a prefix"""
+        if len(lst) < len(prefix):
```

Comment:
I am assuming this is for strings.  Why not use startswith()?

Nevermind... I just noticed it's not for strings... maybe add a comment that it's meant for trajectories.

Reply:
Good point, I'll add a comment to clarify.